### PR TITLE
[Factions] Remove from shared memory and simplify

### PR DESCRIPTION
**Pull request by Kinglykrab**
# Notes
- Removes factions from shared memory and moves to zone based storage of repositories and changes the NPC `faction_list` to also use repositories.
- This affects NPC Factions and Faction Associations.

# Images
## Multi-Faction NPC
image
## Single Faction NPC
image

### CODE REVIEW

**Comment by noudess** [comment]
Give me a couple days on this.  I'll look at it in more detail.

File: zone/worldserver.cpp [hunk 1]
```diff
@@ -2114,6 +2114,24 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
...
+		break;
+	}
+	case ServerOP_ReloadFactionAssociations:
+	{
+		if (zone && zone->IsLoaded()) {
+			zone->SendReloadMessage("Faction Associations");
+			zone->ReloadFactionAssociations();
+		}
+
+		break;
```
Comment (Akkadius) [inline zone/worldserver.cpp:2133]: We spoke about this but I think factions should all be reloaded together (all tables and types)

**Comment by Akkadius** [comment]
Is this ready for review yet, is it fully functional?

**Comment by noudess** [comment]
> Is this ready for review yet, is it fully functional?

Give me another couple of days, been kinda busy,

**Comment by noudess** [comment]
So I am about to visually review again.  Things I'd like to see tested, but myself might need to understand how:

- a kill using mackals new factions splash.  I will look into how,  I think you have to edit npc types and use his new faction_amount field.  Will verify
- a task completed where the task uses mackals new system
- some kind of test that proves mobs assist other factions as per the npc_faction_entries that do not have a faction hit, but only provide assist information.

File: zone/zone_npc_factions.cpp
Comment (noudess) [inline zone/zone_npc_factions.cpp:1]: I'd like to see places where an npc_faction_id is used not to call it a faction_id but rather an npc_faction_id.  A faction_id refers to factions in faction_list, which are in game factions.  An npc_faction_id is an id which is an umbrella under which all faction type activities, primary_faction, hits, behavior, etc is bundled under.

**Comment by noudess** [comment]
I need to take this opportunity to learn how to comment inline on these PRs.  I pointed out several things in my bulk comment above, but I will try and disperse these if @Kinglykrab or someone else can help me learn how.

Factions are not that bad once you understand them.  I think my comments about clearly spelling out what variablkes are npc_faction_ids (these maybe need a new handle, like npc_faction_info_id) as they indicate an umbrella of data, and which variables are actually faction_ids is very important.

The mackal splash faction code also has those names intermixed, and he even overloaded arguments to mean one thing in his system and another thing in the original system.

Hopefully @Kinglykrab and I can get this PR the way we all want it.  Once it's tested and stable, I'll go in and fix any of the remaining confusin with better naming and comments.

Overall, I'd prefer if faction changes were deferred to me when possible.  That said, I would not have been capable of what @Kinglykrab did here, as I am ignorant to the repositories work and much of how shared memory works.  So this PR needed to be collaborative, so its on target.

**Comment by noudess** [comment]
Another important note.  The actual factions themselves, while mostly static, are loaded once with content_db.LoadFactionData().

I think you want to add content_db.LoadFactionData(); to ServerOP_ReloadFactions.

**Comment by noudess** [comment]
We have a sub PR in process between @Kinglykrab  and I.  This is good work in progress.

File: zone/client.cpp [hunk 2]
```diff
@@ -8213,33 +8229,92 @@ void Client::CashReward(uint32 copper, uint32 silver, uint32 gold, uint32 platin
...
 
-	// now hit them in order
-	for (int i = 0; i < MAX_FACTION_ASSOC; ++i) {
-		if (faction_assoc->hits[i].id <= 0) // we don't allow later entries
-			break;
-		if (faction_assoc->hits[i].multiplier == 0.0f) {
-			LogFaction("Bad association multiplier for ID {} entry {}", id, i + 1);
-			continue;
-		}
+	std::vector<int> faction_ids = {
```
Comment (KimLS) [inline zone/client.cpp:8241]: This section is absurdly less readable than the previous code.  

My suggestion is to keep the loop and have a convivence function that returns the correct structure field by index instead.

Comment (noudess) [reply zone/client.cpp:8241]: Can we hold on changes to this until it's merged and retested?  Ive got all the spots I can think of tested and fixes in a [PR ](https://github.com/EQEmu/Server/pull/4039)waiting to be merged, and that includes changes to this function.  I'd like to get this sub-PR merged and then pull the whole repo down so I can test in on my entire server for a day or two before any releases are made.

**Comment by noudess** [comment]
My [PR ](https://github.com/EQEmu/Server/pull/4039)to this branch has tested and fixed all issues I can think of.

- Normal kills
- Kills using mackals splash faction (PR adds this)
- Task faction for both normal and splash (PR adds ths)
- Quest factions for splash (PR adds ths)
- Added code to handle when zone is repopped and some of the vector of factions Loaded exist and others dont

We rooted out bugs in various areas.

I have NOT tested the #reload functionality itself.  I defer that to @Kinglykrab after my PR is merged.  But once this PR is merged in I will move Rolath to current code and regression test that as well as everything else.  Wife and I have a suite of tests we are running everytime I make a fix.

File: zone/zone_npc_factions.cpp
Comment (noudess) [reply zone/zone_npc_factions.cpp:1]: Fixed by my pending [PR  on this branch.

**Review by noudess — APPROVED** [review]
Ready for prime time.  I'll update Rolath and regression immediately after merge,

File: zone/client.cpp (hunk 2 above)
Comment (KimLS) [reply zone/client.cpp:8241]: I think in the future the answer should be "no", I doubt anyone will go back and make the changes now that the merge is complete.

Comment (KimLS) [reply zone/client.cpp:8241]: It is not a big enough deal to revert it or anything so marking as resolved. It's annoying but NBD.

**Pull request merged**